File: server/src/services/ollama.js
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CONFIG_PATH = path.resolve(__dirname, "../../config/llm.yaml");

function parseSimpleYaml(text) {
  const out = {};
  for (const line of String(text || "").split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const m = /^([A-Za-z0-9_]+)\s*:\s*(.+)$/.exec(trimmed);
    if (!m) continue;
    let val = m[2].trim();
    if (val === "true") val = true;
    else if (val === "false") val = false;
    else if (/^\d+$/.test(val)) val = Number(val);
    else if (
      (val.startsWith('"') && val.endsWith('"')) ||
      (val.startsWith("'") && val.endsWith("'"))
    ) {
      val = val.slice(1, -1);
    }
    out[m[1]] = val;
  }
  return out;
}

export function loadLlmConfig() {
  const defaults = {
    enabled: false,
    provider: "ollama",
    model: "llama3.2",
    base_url: "http://localhost:11434",
    timeout_seconds: 25,
  };
  if (!fs.existsSync(CONFIG_PATH)) return defaults;
  try {
    const parsed = parseSimpleYaml(fs.readFileSync(CONFIG_PATH, "utf8"));
    return { ...defaults, ...parsed };
  } catch {
    return defaults;
  }
}

export async function checkOllamaStatus() {
  const config = loadLlmConfig();
  const model = config.model || "llama3.2";
  const baseUrl = String(config.base_url || "http://localhost:11434").replace(/\/$/, "");
  const result = {
    enabled: Boolean(config.enabled),
    model,
    base_url: baseUrl,
    server_ok: false,
    model_ready: false,
    message: "Ollama : vérification…",
  };
  if (!result.enabled) {
    result.message = "Ollama désactivé (config/llm.yaml)";
    return result;
  }
  try {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), 2500);
    const res = await fetch(`${baseUrl}/api/tags`, { signal: ctrl.signal });
    clearTimeout(timer);
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    const body = await res.json();
    result.server_ok = true;
    const installed = new Set(
      (body.models || []).map((m) => String(m.name || "").split(":")[0]),
    );
    result.model_ready = installed.has(model);
    result.message = result.model_ready
      ? `Ollama OK — ${model}`
      : `Modèle absent — ollama pull ${model}`;
  } catch {
    result.message = "Ollama inaccessible — lance ollama serve";
  }
  return result;
}

/**
 * Appel Ollama /api/chat format JSON.
 * @returns {object} parsed JSON
 */
export async function chatJson(messages, { temperature = 0.2, timeoutSeconds } = {}) {
  const config = loadLlmConfig();
  if (!config.enabled) {
    throw new Error("Ollama désactivé");
  }
  const baseUrl = String(config.base_url || "http://localhost:11434").replace(/\/$/, "");
  const timeoutMs =
    Math.max(5, Number(timeoutSeconds ?? config.timeout_seconds) || 25) * 1000;
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(`${baseUrl}/api/chat`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      signal: ctrl.signal,
      body: JSON.stringify({
        model: config.model || "llama3.2",
        messages,
        stream: false,
        format: "json",
        options: { temperature },
      }),
    });
    if (!res.ok) {
      throw new Error(`Ollama HTTP ${res.status}`);
    }
    const body = await res.json();
    const raw = String(body?.message?.content || "").trim();
    try {
      return parseJsonLoose(raw);
    } catch {
      throw new Error("Réponse Ollama JSON invalide");
    }
  } catch (err) {
    if (err?.name === "AbortError") {
      throw new Error("Ollama timeout");
    }
    throw err;
  } finally {
    clearTimeout(timer);
  }
}

function parseJsonLoose(raw) {
  let text = raw.trim();
  if (text.startsWith("```")) {
    text = text.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
  }
  return JSON.parse(text);
}

File: server/app/services/ollama.py
"""Client Ollama + chargement config/llm.yaml."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

import httpx
import yaml

from app.config import settings


def load_llm_config() -> dict[str, Any]:
    defaults: dict[str, Any] = {
        "enabled": False,
        "provider": "ollama",
        "model": "llama3.2",
        "base_url": "http://localhost:11434",
        "timeout_seconds": 25,
    }
    config_path: Path = settings.llm_config_path
    if not config_path.exists():
        return defaults
    try:
        parsed = yaml.safe_load(config_path.read_text(encoding="utf-8")) or {}
        if not isinstance(parsed, dict):
            return defaults
        return {**defaults, **parsed}
    except (OSError, yaml.YAMLError):
        return defaults


def llm_is_usable(status: dict[str, Any] | None) -> bool:
    if not status:
        return False
    return bool(status.get("enabled") and status.get("server_ok") and status.get("model_ready"))


def _model_is_installed(model: str, names: list[str]) -> bool:
    wanted = (model or "").strip()
    if not wanted:
        return False
    if wanted in names:
        return True
    base = wanted.split(":")[0]
    return any(n == base or n.split(":")[0] == base for n in names)


async def check_ollama_status() -> dict[str, Any]:
    config = load_llm_config()
    model = str(config.get("model") or "llama3.2")
    base_url = str(config.get("base_url") or "http://localhost:11434").rstrip("/")
    result: dict[str, Any] = {
        "enabled": bool(config.get("enabled")),
        "model": model,
        "base_url": base_url,
        "server_ok": False,
        "model_ready": False,
        "message": "Ollama : vérification…",
    }
    if not result["enabled"]:
        result["message"] = "Ollama désactivé (config/llm.yaml)"
        return result
    try:
        async with httpx.AsyncClient(timeout=2.5) as client:
            res = await client.get(f"{base_url}/api/tags")
            res.raise_for_status()
            body = res.json()
        result["server_ok"] = True
        names = [str(m.get("name") or "") for m in (body.get("models") or [])]
        result["model_ready"] = _model_is_installed(model, names)
        result["message"] = (
            f"Ollama OK — {model}"
            if result["model_ready"]
            else f"Modèle absent — ollama pull {model}"
        )
    except httpx.ConnectError:
        result["message"] = "Ollama inaccessible — lance ollama serve"
    except httpx.TimeoutException:
        result["message"] = "Ollama timeout au status — lance ollama serve"
    except Exception:
        result["message"] = "Ollama inaccessible — lance ollama serve"
    return result


async def chat_json(
    messages: list[dict[str, str]],
    temperature: float = 0.2,
    timeout_seconds: float | int | None = None,
) -> dict[str, Any]:
    """Appel Ollama /api/chat format JSON."""
    config = load_llm_config()
    if not config.get("enabled"):
        raise RuntimeError("Ollama désactivé")

    base_url = str(config.get("base_url") or "http://localhost:11434").rstrip("/")
    timeout_val = timeout_seconds if timeout_seconds is not None else config.get("timeout_seconds")
    try:
        timeout_s = max(5, float(timeout_val or 25))
    except (TypeError, ValueError):
        timeout_s = 25.0

    try:
        async with httpx.AsyncClient(timeout=timeout_s) as client:
            res = await client.post(
                f"{base_url}/api/chat",
                json={
                    "model": config.get("model") or "llama3.2",
                    "messages": messages,
                    "stream": False,
                    "format": "json",
                    "options": {"temperature": temperature},
                },
            )
    except httpx.ConnectError as exc:
        raise RuntimeError("Ollama inaccessible — lance ollama serve") from exc
    except httpx.TimeoutException as exc:
        raise RuntimeError(f"Ollama timeout ({timeout_s:.0f}s)") from exc
    except httpx.HTTPError as exc:
        raise RuntimeError(f"Ollama réseau : {exc}") from exc

    if res.status_code >= 400:
        snippet = (res.text or "").strip().replace("\n", " ")[:180]
        raise RuntimeError(f"Ollama HTTP {res.status_code}{f' — {snippet}' if snippet else ''}")

    try:
        body = res.json()
    except json.JSONDecodeError as exc:
        raise RuntimeError("Réponse Ollama non JSON") from exc

    raw = str((body.get("message") or {}).get("content") or "").strip()
    try:
        return _parse_json_loose(raw)
    except (json.JSONDecodeError, TypeError, ValueError) as exc:
        raise RuntimeError("Réponse Ollama JSON invalide") from exc


def _parse_json_loose(raw: str) -> dict[str, Any]:
    text = raw.strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text, flags=re.IGNORECASE)
        text = re.sub(r"\s*```$", "", text)
    parsed = json.loads(text)
    if not isinstance(parsed, dict):
        raise ValueError("JSON root must be an object")
    return parsed
